### ISSUE TITLE
YOLOE (and maybe others) uses wrong device even when specified

### ISSUE THREAD

**Issue by Jordan-Pierce** [ultralytics/ultralytics#21664]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

Hi, looking into YOLOE and found that despite setting the model to use `cpu`, because I have a CUDA gpu, it will force switch over to device `0` when making predictions:

<img width="535" height="221" alt="Image" src="https://github.com/user-attachments/assets/c9f091aa-fa45-44b9-bd2c-c6945c82331e" />

<img width="1047" height="735" alt="Image" src="https://github.com/user-attachments/assets/57c305b5-1571-4e4a-997d-7b74d0855f54" />

### Environment

```
Ultralytics 8.3.175  Python-3.10.18 torch-2.8.0+cu126 CUDA:0 (NVIDIA GeForce RTX 3090 Ti, 24564MiB)
Setup complete  (24 CPUs, 127.7 GB RAM, 220.7/1862.2 GB disk)

OS                  Windows-10-10.0.26100-SP0
Environment         Windows
Python              3.10.18
Install             pip
RAM                 127.75 GB
Disk                220.7/1862.2 GB
CPU                 12th Gen Intel Core(TM) i9-12900K
CPU count           24
GPU                 NVIDIA GeForce RTX 3090 Ti, 24564MiB
GPU count           1
CUDA                12.6

numpy                2.2.6>=1.23.0
matplotlib           3.10.5>=3.3.0
opencv-python        4.12.0.88>=4.6.0
pillow               11.3.0>=7.1.2
pyyaml               6.0.2>=5.3.1
requests             2.32.4>=2.23.0
scipy                1.15.3>=1.4.1
torch                2.8.0+cu126>=1.8.0
torch                2.8.0+cu126!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision          0.23.0+cu126>=0.9.0
tqdm                 4.67.1>=4.64.0
psutil               7.0.0
py-cpuinfo           9.0.0
pandas               2.3.1>=1.1.4
ultralytics-thop     2.0.15>=2.0.0
```

### Minimal Reproducible Example

See screenshots above

### Additional

This doesn't seem to happen with other models

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Jordan-Pierce, thanks for opening this issue and for the detailed environment info 🚀 This is an automated response — an Ultralytics engineer will review and assist you here soon.

Please check the resources below while we investigate:

- Start with the Docs for usage guidance: [Docs](https://docs.ultralytics.com/), including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples.
- If this is a 🐛 Bug Report (it appears to be), please provide a complete [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) rather than screenshots so we can reproduce locally. Include:
  - Exact code snippet showing model loading and inference, including any device arguments you pass
  - Model variant and weights used (e.g., YOLOE and specific checkpoint)
  - Console logs with INFO/DEBUG output
  - Any custom settings or overrides used
- If this is a custom training ❓ Question, please share dataset samples and training logs, and ensure you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, visit [Discourse](https://community.ultralytics.com/)
- For community threads, see our [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

First, please upgrade to the latest ultralytics release in a clean environment to verify whether the issue persists. Ensure all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) are satisfied in a Python>=3.8 environment with PyTorch>=1.8:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date verified environments (with dependencies like CUDA/CUDNN, Python, and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To move forward, could you please share a minimal reproducible example (code-only), the exact YOLOE weights used, and the full console logs showing the device selection behavior? This will help us pinpoint whether the predictor, model loading, or a specific argument path is forcing CUDA despite a CPU request. Thanks! 🙏

**Comment by RizwanMunawar** [comment]
Thanks for reporting this, @Jordan-Pierce. I can reproduce the issue and have opened a PR #21670 with the fix. Once it’s merged, the problem will be resolved.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in PR #21670 by @RizwanMunawar with contributions from @Laughing-q and @glenn-jocher.

Summary of what changed:
- Device and half flags are now properly propagated through Model.predict to internal predictors, ensuring consistent GPU/AMP behavior.
- YOLOE predictor builds visual prompt tensors once and casts to float16 when running in half, otherwise float32, reducing dtype mismatches.
- Attention block simplifies softmax casting to operate in the current dtype for improved stability and consistency.

How to test the fix:
- Install the latest main branch:
  pip install -U git+https://github.com/ultralytics/ultralytics.git@main
- Alternatively, wait for the next official PyPI release and update when available:
  pip install -U ultralytics
- After updating, verify your workflow with explicit device and/or half settings (e.g., device=0, half=True) and check that prompts and attention behave consistently without dtype or device-related errors.

Please let us know if the merged changes resolve the issue. Thanks for reporting this, and feel free to share logs or a minimal repro if the problem persists. 🙏
